paimon-web-server/src/main/java/org/apache/paimon/web/server/controller/UserController.java

```diff
@@ -41,19 +53,76 @@ public class UserController {
     @Autowired private UserService userService;
 
     /**
-     * get user by id.
+     * Get user by id.
      *
      * @param id user-id
-     * @return {@link R} with {@link User}
+     * @return {@link R} with {@link UserVO}
      */
     @SaCheckPermission("system:user:query")
     @GetMapping("/{id}")
-    public R<User> getUserById(@PathVariable("id") Long id) {
-        User user = userService.getById(id);
+    public R<UserVO> getUser(@PathVariable("id") Integer id) {
+        UserVO user = userService.getUserById(id);
         if (user == null) {
             return R.failed(USER_NOT_EXIST);
         }
-        user.setPassword(null);
         return R.succeed(user);
     }
+
+    /**
+     * Get user views with pagination.
+     *
+     * @param user filter conditions
+     * @return paginated user view objects
+     */
+    @SaCheckPermission("system:user:list")
+    @GetMapping("/list")
+    public PageR<UserVO> listUsers(User user) {
+        IPage<User> page = PageSupport.startPage();
+        List<UserVO> list = userService.listUsers(page, user);
+        return PageR.<UserVO>builder().success(true).total(page.getTotal()).data(list).build();
+    }
+
+    /**
+     * Add a new user.
+     *
+     * @param user the user to be added, must not be null
+     * @return a {@code R<Void>} response indicating success or failure
+     */
+    @SaCheckPermission("system:user:add")
+    @PostMapping
+    public R<Void> add(@Validated @RequestBody User user) {
+        if (!userService.checkUserNameUnique(user)) {
+            return R.failed(Status.USER_NAME_ALREADY_EXISTS, user.getUsername());
+        }
+
+        return userService.insertUser(user) > 0 ? R.succeed() : R.failed();
+    }
+
+    /**
+     * Update an existing user's details.
+     *
+     * @param user the user with updated details, must not be null
+     * @return a {@code R<Void>} response indicating success or failure
+     */
+    @SaCheckPermission("system:user:update")
+    @PutMapping
+    public R<Void> update(@Validated @RequestBody User user) {
+        if (!userService.checkUserNameUnique(user)) {
+            return R.failed(Status.USER_NAME_ALREADY_EXISTS, user.getUsername());
+        }
+
+        return userService.updateUser(user) > 0 ? R.succeed() : R.failed();
+    }
+
+    /**
+     * Delete one or more users by user ID.
+     *
+     * @param userIds an array of user IDs to be deleted
+     * @return a {@code R<Void>} response indicating success or failure
+     */
+    @SaCheckPermission("system:user:delete")
+    @DeleteMapping("/{userIds}")
+    public R<Void> delete(@PathVariable Integer[] userIds) {
+        return userService.deleteUserByIds(userIds) > 0 ? R.succeed() : R.failed();
```

Comment:
What's the behavior that the same user is multiple deleted?

Reply:
It has been modified. This should not be roleId.

Reply:
> What's the behavior that the same user is multiple deleted?

I think there should be no situation where a User is deleted multiple times, because this is transmitted from the front end. If a User is deleted, it will not be displayed in the User list, so the front end cannot pass the deleted UserId to the back end. Endless.

Reply:
@s7monk, when the same user is multiple deleted, we should return succeed instead of failed. The backend interface has nothing with frontend.

Reply:
done.

Reply:
> What's the behavior that the same user is multiple deleted?

In my test, I tested passing in the same userId, and the result was success.